File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.java
class Solution {
    public int minReorder(int n, int[][] connections) {
        List<List<int[]>> graph = new ArrayList<>();
        for(int i = 0 ; i<n ; i++){
            graph.add(new ArrayList<>());
        } 
        for(int[] ni : connections){
            int u = ni[0] , v = ni[1];
            graph.get(u).add(new int[]{v,1});
            graph.get(v).add(new int[]{u,0});
        }
        boolean[] v = new boolean[n];
        return dfs(0,graph,v);
    }

    int dfs(int n ,List<List<int[]>> graph,boolean[] v){
        int c = 0;
        v[n] = true;
        for(int[] ne : graph.get(n)){
            int e = ne[0] ,  needC = ne[1];
            if(!v[e]){
                c += needC + dfs(e,graph,v);
            }
        }
        return c;
    }
}

File: 1576-reorder-routes-to-make-all-paths-lead-to-the-city-zero/reorder-routes-to-make-all-paths-lead-to-the-city-zero.py
class Solution:
    def minReorder(self, n: int, connections: List[List[int]]) -> int:
        graph = defaultdict(list)
        directed = set()

        for a,b in connections:
            graph[a].append(b)
            graph[b].append(a)
            directed.add((a,b))

        v = [False] * n

        def dfs(node : int) -> int:
            v[node] = True
            changes = 0

            for neighbour in graph[node]:
                if not v[neighbour]:
                    if (node , neighbour) in directed:
                        changes += 1

                    changes += dfs(neighbour)
            return changes

        return dfs(0)



        